### PR TITLE
Update to ol@7

### PR DESCRIPTION
**Pull request by oterral**
This PR contains a massive update of openlayers v7.

 index.php and weather.php have been updated . 
All layers works as before.
Trip planner tool and permalink also.

Overview map has been removed, because it is pretty useless and it didn 't work on original website.

Ready for heavy testing.

### CODE REVIEW

**Comment by stevo01** [comment]
Hello,
I startet a test instance for a review of your changes. 
url:https://test1-alpha.openseamap.org.
 
note: I got following error from webbrowser when I try to open the URL:
Access to image at 'https://t2.openseamap.org/tile/9/271/163.png' from origin 'https://test1-alpha.openseamap.org' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
t2.openseamap.org/tile/9/271/163.png:1          Failed to load resource: net::ERR_FAILED

**Comment by oterral** [comment]
@stevo01   Thx. But the urls you gave me, it doesn't seem to have the code of this pull request. It's still the current master branch. 

It should look like this, see zoom buttons ad zoom bar :
image

**Comment by stevo01** [comment]
@oterral : The webserver serving https://test1-alpha.openseamap.org uses the following repository and commit ID:
```
bash-4.3# git remote -v
origin  https://github.com/oterral/online_chart.git (fetch)
origin  https://github.com/oterral/online_chart.git (push)
bash-4.3# git status
HEAD detached at 5c5b025
nothing to commit, working tree clean
```
and I assume thats the version that you used for the pull request. Let me know if I mixed up something.

**Comment by stevo01** [comment]
@oterral : After bugfix in dockerfile your version is visible. 
https://github.com/oterral/online_chart.git  / commit 5c5b025

note: beside that I changed the config file "online_chart.ini" for usage of OpenStreatMap Tileserver (https://tile.openstreetmap.org/). My webbrowser shows still a CORS error when OpenSeaMap tileserver is used (https://t2.openseamap.org). Can you crosscheck that on your testsystem?

**Comment by oterral** [comment]
@stevo01   For the  CORS issue , the tile backend must send the image with the following  header in the response: 

access-control-allow-methods: GET, OPTIONS
access-control-allow-origin: *

These headera are there when using https://tile.openstreetmap.org/ and https://tiles.openseamap.org/seamark but not when using https://t2.openseamap.org/

**Comment by stevo01** [comment]
> @stevo01 For the CORS issue , the tile backend must send the image with the following header in the response:
> 
> access-control-allow-methods: GET, OPTIONS access-control-allow-origin: *
> 
> These headera are there when using https://tile.openstreetmap.org/ and https://tiles.openseamap.org/seamark but not when using https://t2.openseamap.org/

Thanks for the feedback. I will check it.

**Comment by stevo01** [comment]
>

**Comment by stevo01** [comment]
@oterral: I added parameter "crossOrigin: null" in sources from test server to fix the issue with cors. 
I assume this disables the CORS feature. Can you take over that setting in your pull request? From my point of view we can decide later if CORS protection is needed and how to realise it on tile server. 

image

File: index.php [hunk 1]
```diff
@@ -695,7 +695,7 @@ function updateCheckboxAndCookie(layer) {
                 //                                         { layerId      : 1,
                 //                                           wrapDateLine : true
                 //                                         });
-                const osmUrl = OsmTileServer == "BRAVO" ? 'https://t2.openseamap.org/tile/{z}/{x}/{y}.png' : 'https://tile.openstreetmap.org/{z}/{x}/{y}.png';
+                const osmUrl = OsmTileServer !== "BRAVO" ? 'https://t2.openseamap.org/tile/{z}/{x}/{y}.png' : 'https://tile.openstreetmap.org/{z}/{x}/{y}.png';
```
Comment (wschildbach) [inline index.php:698]: This change seems unrelated to the rose, and inverts the sense of the switch for selecting the back-up tileserver.

File: index.php [hunk 2]
```diff
@@ -519,19 +525,19 @@ function setWikiLayer(visible){
...
+            //             url: 'api/proxy-wikipedia.php?',
+            //       params: {
+            //           'LANG' : language,
+            //           'thumbs' : displayThumbs
+            //       },
+            //       format: new OpenLayers.Format.KML({
+            //           extractStyles: true,
+            //           extractAttributes: true
+            //       })
+            //   });
```
Comment (wschildbach) [inline index.php]: Did you move all the commented functionality elsewhere or is this still a to-do?

File: index.php [hunk 3]
```diff
@@ -778,19 +784,6 @@ function drawmap() {
...
-                //     },
-                //     format: new OpenLayers.Format.KML({
-                //         extractStyles: true,
-                //         extractAttributes: true
-                //     })
-                // });
-
                 // Select feature ---------------------------------------------------------------------------------------------------------
                 // (only one SelectFeature per map is allowed)
                 // TODO oli
```
Comment (wschildbach) [inline index.php:650]: Is this still a to-do?

**Comment by wschildbach** [comment]
I started reviewing the code, but since I don't have a laptop available at the moment I stopped again for now... will take up again end of the week.
Is this ready for review or still WIP?

File: index.php (hunk 2 above)
Comment (oterral) [reply index.php]: This commented code must be removed. The update  of the layer when `wikipediaThumbs` variable changes is implemented is in `layer_wikipedia` layer definition and triggered via the next line "`layer_wikipedia.getSource().refresh()`

File: index.php (hunk 3 above)
Comment (oterral) [reply index.php:650]: No the Select control is not needed anymore

Comment (oterral) [reply index.php:650]: basically all the TODO are done I think . I just waited the tests from Markus to remove theme

File: index.php (hunk 1 above)
Comment (oterral) [reply index.php:698]: Yes t2 server had CORS issue so I used the other one, but I will put back the old comparaison

**Comment by oterral** [comment]
@wschildbach   maybe it's better to review now before I remove/move some  code and files. After, it will be difficult to see what has changed.

File: api/help/help-josm_en.php [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<iframe src="https://wiki.openseamap.org/wiki/JOSM_and_Plugin" width="540" height="670" frameborder="0">
+<p>Your browser can not show embedded frames:
+You may view the embedded site over the following
```
Comment (wschildbach) [inline api/help/help-josm_en.php:3]: Minor: I think "through the following" would be a better wording. And you can use "link" instead of "reference".

File: api/help/help-online-editor_en.php [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<iframe src="https://wiki.openseamap.org/wiki/Online-Editor" width="540" height="670" frameborder="0">
+<p>Your browser can not show embedded frames:
+You may view the embedded site over the following
```
Comment (wschildbach) [inline api/help/help-online-editor_en.php:3]: Minor: I think "through the following" would be a better wording. And you can use "link" instead of "reference".

Comment (wschildbach) [reply api/help/help-online-editor_en.php:3]: Same comment for all the following _en pages...

File: api/help/help-tidal-scale_de.php [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<iframe src="https://wiki.openseamap.org/wiki/De:Pegel" width="540" height="670" frameborder="0">
+  <p>Ihr Browser kann leider keine eingebetteten Frames anzeigen: 
+  Sie können die eingebettete Seite über den folgenden 
+  Verweis aufrufen: 
+  https://wiki.openseamap.org/index.php?title=De:Pegel&printable=yes
```
Comment (wschildbach) [inline api/help/help-tidal-scale_de.php:5]: This is the only link with &printable=yes... is this by intention?

File: javascript/map_utils.js [hunk 7]
```diff
@@ -106,263 +39,228 @@ function lonLatToMercator(ll) {
...
-            break;
-        case '°':
-            if (fill === '#') {
-                throw 'missing format specifier';
-            }
-            /* If decimal places are to be rendered, use the full number.
+    return s;
+  }
+
+  // OL2 has a bug (?) where sometimes the coordinates go beyond +-180
```
Comment (wschildbach) [inline javascript/map_utils.js:56]: Is this code still necessary or appropriate? It seems to refer to OL2 only.

File: javascript/permalink.js [hunk 8]
```diff
@@ -1,119 +1,52 @@
...
+    this.timeout = null;
+
+    this.aElement_ = document.createElement("a");
+    this.aElement_.innerHTML = "Permalink";
+    this.element.className = "ol-permalink";
+    this.element.appendChild(this.aElement_);
+  }
+
+  render() {
+    // We use a timeout to avoid unnecessary updates.
```
Comment (wschildbach) [inline javascript/permalink.js:46]: The way I understand the code is that it does not really remove the updates but moves them past the phase where the map would be scrolling. It would not actually cancel redundant or outdated updates.
To be clear, I don't see this as a problem, just trying to understand what this does.

File: index.php [hunk 9]
```diff
@@ -744,229 +472,811 @@ function addSearchResults(xmlHttp) {
...
+
                 //GebcoDepth
-                layer_gebco_deeps_gwc = new OpenLayers.Layer.WMS("gebco_2021", "https://depth.openseamap.org/geoserver/gwc/service/wms",
-                    {layers: "gebco2021:gebco_2021", format:"image/png"},
-                    { layerId: 6, isBaseLayer: false, visibility: false, opacity: 0.8});
+                // old definition
+                // layer_gebco_deeps_gwc = new OpenLayers.Layer.WMS("gebco_2021", "https://depth.openseamap.org/geoserver/gwc/service/wms",
+                //     {layers: "gebco2021:gebco_2021", format:"image/png"},
+                //     { layerId: 6, isBaseLayer: false, visibility: false, opacity: 0.8});
+                layer_gebco_deeps_gwc = new ol.layer.Tile({
```
Comment (wschildbach) [inline index.php:674]: The Gebco layer seems broken. Best I can determine, the new code leads to parameters WIDTH and HEIGTH set to values that are not 256 (the map server seems to only support 256). On my machine, WIDHT=HEIGHT=307.

**Comment by wschildbach** [comment]
I've been through the code once now. Great you were able to do a lot of cleanup and refactoring of the code! Thank you a million, this really makes the code base better and more stable.

I left some minor remarks for help texts and such. The compass rose has a bug (see my comment there with the culprit) and the gebco ("Marine profile") layer does not work (see again my comment there).

File: api/help/help-josm_en.php (hunk 4 above)
Comment (oterral) [reply api/help/help-josm_en.php]: Fixed

File: api/help/help-online-editor_en.php (hunk 5 above)
Comment (oterral) [reply api/help/help-online-editor_en.php]: fixed

File: api/help/help-tidal-scale_de.php (hunk 6 above)
Comment (oterral) [reply api/help/help-tidal-scale_de.php:5]: No idea it was like that before

File: javascript/map_utils.js (hunk 7 above)
Comment (oterral) [reply javascript/map_utils.js:56]: I adapted the comment but I will not change the code. Too scary :D and not time to check.

File: javascript/permalink.js (hunk 8 above)
Comment (oterral) [reply javascript/permalink.js:46]: The timeout makes the HTML updates  wait that  the map is in idle state for 500ms then update the permalink with the last state. It just avoid having the html rerendered dozen times instead of one.

File: index.php (hunk 9 above)
Comment (oterral) [reply index.php:674]: Works for me . It seems it's broken only when the device has pxelRatio > 1.

Comment (oterral) [reply index.php:674]: Thanks for pointing this out , I have deactivated the hidpi functionnaility for all TileWMS source.

File: javascript/mag_deviation.js [hunk 10]
```diff
@@ -21,51 +21,80 @@
...
 
-    if (geoMag == undefined) {
-        function initModel(data) {
-            var wmm = cof2Obj(data);
-            geoMag = geoMagFactory(wmm);
-            setMagdev($(this)[0]);
-        }
+  const [lon0, lat0] = ol.proj.toLonLat([bounds[0], bounds[1]]);
+  const [lon1, lat1] = ol.proj.toLonLat([bounds[2], bounds[3]]);
+  var b = lat0.toFixed(5);
```
[the remark this answers is not in the document]
Comment (oterral) [reply javascript/mag_deviation.js]: Good found. I fixed it and remove hte toFixed(5) which, I guess, was there to avoid requesting an url too long (old browser limitation maybe). but it seems to work now.

**Comment by oterral** [comment]
@wschildbach   All is fixed . Thanks again for the review. You can check the 4 last commit to see what I have changed
